### DISCUSSION TITLE
Banging my head against this

### DISCUSSION THREAD

**Discussion by Tsuiban** [slint-ui/slint#4170] [General] [answered]
Ok, trying to write something with slint in Rust, and utterly frustrated with it.

1. It seems that I need to create a weak reference for EACH callback, so multiple variables instead of one single one.
2. I have a few LineEdit fields that, supposedly, take their values from a property and, theoretically, should be putting edited values back into those properties - and yet that doesn't seem to be happening.  The properties remain blank.

so in my main code I have something resembling:

main.rs:
```rust
...
let ui = AppWindow::new();
let ui_1 = ui.as_weak();
let ui_2 = ui.as_weak();
...
ui.on_request_name_accepted(move || { let ui = ui_1.unwrap(); let model = ui.get_model_property(); println!("{model:?}"); });
ui.on_request_start_accepted(move || { let ui = ui_2.unwrap(); let model = ui.get_model_property(); println!("{model:?}"); });
ui.on_request_end_accepted(move || { let ui = ui_3.unwrap().let model = ui.get_model_property(); println!("{model:?}"); });
...
```
appwindow.slint:
```rust
struct Model {
    name: string,
    start: string,
    end: string,
}

export component AppWindow inherits Window {
    callback request_name_accepted();
    callback request_start_accepted();
    callback request_end_accepted();
    in-out property<Model> model_property;
...

    LineEdit {
        placeholder-text: "Unnamed";
        text: root.model_property.name;
        accepted =>  { root.request-name-accepted(); }
    }
...
```

The callback in my rust code is being called, but all three fields of model are always "".

**Comment by ogoffart** [comment]
```
        text: root.model_property.name;
```

This is a one way binding. The text is assigned to the model, but not the other way arround.  You need to update it in the accepted, for example

```
accepted =>  { 
   model_property.name = self.text;
   root.request-name-accepted();
 }

```

**Comment by bjorn** [comment, marked as answer]
> 1. It seems that I need to create a weak reference for EACH callback, so multiple variables instead of one single one.

I see this mostly as a shortcoming of Rust, since it appears lambdas can only use references (the default) or move values (when writing `move ||`, whereas C++ provides `[=] {` to create a lambda based on copies. To avoid the annoying creation of variables like `ui_1`, `ui_2`, etc. I have gotten into the habit of using this nice Rust feature (not sure what to call it):

```rust
ui.on_request_name_accepted({ let ui = ui.as_weak(); move || { let model = ui.unwrap().get_model_property(); println!("{model:?}"); }});
ui.on_request_start_accepted({ let ui = ui.as_weak(); move || { let model = ui.unwrap().get_model_property(); println!("{model:?}"); }});
ui.on_request_end_accepted({ let ui = ui.as_weak(); move || { let model = ui.unwrap().get_model_property(); println!("{model:?}"); }});
```

It narrows the scope of the weak `ui` variable, which I think is more readable. Almost like the C++ version, which would be `[ui = slint::ComponentWeakHandle(ui)] { ... }` (I think a convenience function would be nice here as well).

It might have been nice if the UI callbacks automatically received a reference to the UI as a parameter, but maybe there was a good reason to omit this.

**Reply by tronical** [reply]
One of the things we've contemplated doing is having support for callbacks were the strong reference is already a mandatory argument of the callback. One issue is that it's not backwards compatible, so we'd have to invent a new scheme or so:

```rust
ui.on_request_name_accepted_with_context(|ui| { ui.do_strong_stuff_here_no_need_for_unwrap() }); 
```

Maybe there's a more elegant way of expressing this in the generated API?

**Reply by bjorn** [reply]
Maybe there could be an option set in the `CompilerConfiguration` through `build.rs` to change the generated code for the callbacks to include such a parameter? Then people can set this option and adjust their code in their own pace, and in Slint 2 the default could be flipped if it turns out to be simply more convenient.

**Reply by Tsuiban** [reply]
While a little cumbersome (but not your fault for the subersome-ness), that's a great idea, Bjorn.  Thank you!

**Reply by tronical** [reply]
`CompilerConfiguration` is a good idea. Perhaps it could also be a kind of `@codegen-pragma(rust:callbacks-with-component-handle)` (or something shorter). But a project wide setting is probably most convenient.

**Reply by ogoffart** [reply]
btw, some implementation work was done before in https://github.com/slint-ui/slint/pull/2035
IIRC we didn't include that because it did not simplify most of our example.  But perhaps it does simplify the actual complex apps

**Reply by bjorn** [reply]
Hmm, it is true that it actually would not help my app much, and I also have my doubts about it being generally useful. This is because the callback could only take a reference to the UI component, which is rarely enough in a complex app.

Usually one needs access to other application state, so this cloning dance will still be required, and since you can put a reference to the UI in your application state, at least my code could not be simplified at all this way. It actually looks like:
```rust
struct App {
    // ...
    ui_weak: slint::Weak<AppWindow>,
}

#[tokio::main]
async fn main() -> Result<()> {
    let mut app = App::new();

    // ...

    let app = Arc::new(Mutex::new(app));
    let facade = ui.global::<Facade>();

    facade.on_ui_index_for_transaction({
        let app = app.clone();

        move |tx_index| {
            let app = app.lock();
            // ...
        }
    });

    // ...

    ui.run()?;
    Ok(())
}
```
(and I need `Arc` instead of `Rc` due to triggering some async tokio code in callbacks, even if I don't need this to be multi-threaded...)
